gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/UserServiceImpl.java

```diff
@@ -287,44 +292,47 @@ public Completable resetPassword(Domain domain, String userId, String password,
-                                                    .onErrorResumeNext(ex -> {
-                                                        if (ex instanceof UserNotFoundException) {
-                                                            // idp user not found, create its account
-                                                            user.setPassword(password);
-                                                            return userProvider.create(convert(user));
-                                                        }
-                                                        return Single.error(ex);
-                                                    });
-                                        })
-                                        .flatMap(idpUser -> {
-                                            // update 'users' collection for management and audit purpose
-                                            // if user was in pre-registration mode, end the registration process
-                                            if (user.isPreRegistration()) {
-                                                user.setRegistrationCompleted(true);
-                                                user.setEnabled(true);
-                                            }
-                                            user.setPassword(null);
-                                            user.setExternalId(idpUser.getId());
-                                            user.setLastPasswordReset(new Date());
-                                            user.setUpdatedAt(new Date());
-                                            return userService.update(user);
-                                        })// after audit, invalidate tokens whatever is the domain or app settings
-                                        // as it is an admin action here, we want to force the user to login
-                                        .flatMap(updatedUser -> Single.defer(() -> tokenService.deleteByUserId(updatedUser.getId())
-                                                .toSingleDefault(updatedUser)
-                                                .onErrorResumeNext(err -> {
-                                                    logger.warn("Tokens not invalidated for user {} due to : {}", userId, err.getMessage());
-                                                    return Single.just(updatedUser);
-                                                })))
-                                        .doOnSuccess(user1 -> auditService.report(AuditBuilder.builder(UserAuditBuilder.class).client(client).principal(principal).type(EventType.USER_PASSWORD_RESET).user(user)))
+                                final Client client = optClient.map(Application::toClient).orElse(new Client());
+                                return passwordHistoryService.addPasswordToHistory(DOMAIN, domain.getId(), user, password , principal, domain.getPasswordSettings())
```

Comment:
We are only using the domain settings, we should use` PasswordSetting.getInstance`